thriftbp/server_middlewares.go

```diff
@@ -337,3 +340,38 @@ func tHeaderProtocol2String(proto thrift.THeaderProtocolID) string {
 		return "binary"
 	}
 }
+
+// RecoverPanic recovers from panics raised in the TProccessorFunction chain, reports them
+// to sentry, and records a metric indicating that the endpoint recovered from a panic.
+func RecoverPanic(name string, next thrift.TProcessorFunction) thrift.TProcessorFunction {
+	return thrift.WrappedTProcessorFunction{
+		Wrapped: func(ctx context.Context, seqId int32, in, out thrift.TProtocol) (ok bool, err thrift.TException) {
+			defer func() {
+				if r := recover(); r != nil {
+					var rErr error
+					if asErr, ok := r.(error); ok {
+						rErr = asErr
+					} else {
+						rErr = fmt.Errorf("panic in %q: %+v", name, r)
+					}
+					log.ErrorWithSentry(
+						ctx,
+						"recovered from panic:",
+						rErr,
+						"endpoint", name,
+					)
```

Comment:
Should this be a logger that is passed in rather than `log.ErrorWithSentry`?

Reply:
As an option maybe but it is the right default.